File: apprefurbished/src/main/java/com/androidcollider/easyfin/home/HomeFragment.java
package com.androidcollider.easyfin.home;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.util.Pair;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.CheckBox;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.androidcollider.easyfin.R;
import com.androidcollider.easyfin.common.app.App;
import com.androidcollider.easyfin.common.events.UpdateFrgHome;
import com.androidcollider.easyfin.common.events.UpdateFrgHomeBalance;
import com.androidcollider.easyfin.common.events.UpdateFrgHomeNewRates;
import com.androidcollider.easyfin.common.managers.chart.setup.ChartSetupManager;
import com.androidcollider.easyfin.common.managers.rates.rates_info.RatesInfoManager;
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager;
import com.androidcollider.easyfin.common.managers.shared_pref.SharedPrefManager;
import com.androidcollider.easyfin.common.managers.ui.dialog.DialogManager;
import com.androidcollider.easyfin.common.ui.MainActivity;
import com.androidcollider.easyfin.common.ui.adapters.SpinIconTextHeadAdapter;
import com.androidcollider.easyfin.common.ui.fragments.common.CommonFragmentWithEvents;
import com.androidcollider.easyfin.common.utils.ChartLargeValueFormatter;
import com.github.mikephil.charting.charts.HorizontalBarChart;
import com.github.mikephil.charting.charts.PieChart;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.PieData;

import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.util.Map;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;

import static butterknife.ButterKnife.findById;

/**
 * @author Ihor Bilous
 */

public class HomeFragment extends CommonFragmentWithEvents implements HomeMVP.View {

    @BindView(R.id.spinMainPeriod)
    Spinner spinPeriod;
    @BindView(R.id.spinMainCurrency)
    Spinner spinBalanceCurrency;
    @BindView(R.id.spinMainChart)
    Spinner spinChartType;
    @BindView(R.id.tvMainStatisticSum)
    TextView tvStatisticSum;
    @BindView(R.id.tvMainSumValue)
    TextView tvBalanceSum;
    @BindView(R.id.tvMainNoData)
    TextView tvNoData;
    @BindView(R.id.tvMainCurrentBalance)
    TextView tvBalance;
    @BindView(R.id.ivMainBalanceSettings)
    ImageView ivBalanceSettings;
    @BindView(R.id.chartHBarMainStatistic)
    HorizontalBarChart chartStatistic;
    @BindView(R.id.chartMainBalance)
    HorizontalBarChart chartBalance;
    @BindView(R.id.chartPieMainStatistic)
    PieChart chartStatisticPie;
    CheckBox chkBoxConvert;
    CheckBox chkBoxShowOnlyIntegers;

    private MaterialDialog balanceSettingsDialog;

    private boolean convert, showOnlyIntegers,

    // we can prevent init spinner issue with subscribing
    // to listeners in onRestoreInstanceState
    spinPeriodNotInitSelectedItemCall,
            spinBalanceCurrencyNotInitSelectedItemCall,
            spinChartTypeNotInitSelectedItemCall;

    @Inject
    RatesInfoManager ratesInfoManager;

    @Inject
    SharedPrefManager sharedPrefManager;

    @Inject
    ResourcesManager resourcesManager;

    @Inject
    ChartSetupManager chartSetupManager;

    @Inject
    DialogManager dialogManager;

    @Inject
    HomeMVP.Presenter presenter;


    @Override
    public int getContentView() {
        return R.layout.frg_home;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((App) getActivity().getApplication()).getComponent().inject(this);
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        initializeViewsAndRes();

        setBalanceCurrencySpinner();
        setStatisticPeriodSpinner();

        presenter.setView(this);
        presenter.loadBalanceAndStatistic(spinPeriod.getSelectedItemPosition() + 1);
    }

    private void initializeViewsAndRes() {
        setupCharts();

        buildBalanceSettingsDialog();

        View balanceSettings = balanceSettingsDialog.getCustomView();

        if (balanceSettings != null) {
            chkBoxConvert = findById(balanceSettings, R.id.checkBoxMainBalanceSettingsConvert);
            chkBoxShowOnlyIntegers = findById(balanceSettings, R.id.checkBoxMainBalanceSettingsShowCents);
        }

        convert = sharedPrefManager.getMainBalanceSettingsConvertCheck();
        showOnlyIntegers = sharedPrefManager.getMainBalanceSettingsShowOnlyIntegersCheck();

        chkBoxConvert.setChecked(convert);
        chkBoxShowOnlyIntegers.setChecked(showOnlyIntegers);

        chkBoxConvert.setOnCheckedChangeListener((compoundButton, b) -> {
            sharedPrefManager.setMainBalanceSettingsConvertCheck(b);
            convert = b;
            setBalance(spinBalanceCurrency.getSelectedItemPosition());

            presenter.updateTransactionStatisticArray(spinBalanceCurrency.getSelectedItemPosition());
            setStatisticSumTV();
            checkStatChartTypeForUpdate();
        });

        chkBoxShowOnlyIntegers.setOnCheckedChangeListener((compoundButton, b) -> {
            sharedPrefManager.setMainBalanceSettingsShowOnlyIntegersCheck(b);
            showOnlyIntegers = b;
            setBalance(spinBalanceCurrency.getSelectedItemPosition());
            checkStatChartTypeForUpdate();
        });

        ratesInfoManager.setupMultiTapListener(tvBalance, getActivity());
    }

    private void buildBalanceSettingsDialog() {
        MainActivity activity = (MainActivity) getActivity();
        if (activity != null) {
            balanceSettingsDialog = dialogManager.buildBalanceSettingsDialog(activity);
        }
    }

    private void setBalanceCurrencySpinner() {
        spinBalanceCurrency.setAdapter(new SpinIconTextHeadAdapter(
                getActivity(),
                R.layout.spin_head_icon_text_main,
                R.id.tvSpinHeadIconTextMain,
                R.id.ivSpinHeadIconTextMain,
                R.layout.spin_drop_icon_text,
                R.id.tvSpinDropIconText,
                R.id.ivSpinDropIconText,
                resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY),
                resourcesManager.getIconArray(ResourcesManager.ICON_FLAGS)));

        spinBalanceCurrency.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                if (spinBalanceCurrencyNotInitSelectedItemCall) {
                    setBalance(i);
                    presenter.updateTransactionStatisticArray(i);
                    setStatisticSumTV();
                    checkStatChartTypeForUpdate();
                    sharedPrefManager.setHomeBalanceCurrencyPos(i);
                } else {
                    spinBalanceCurrencyNotInitSelectedItemCall = true;
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {
            }
        });

        spinBalanceCurrency.setSelection(sharedPrefManager.getHomeBalanceCurrencyPos());
    }

    private void setStatisticPeriodSpinner() {
        ArrayAdapter<?> adapterStatPeriod = ArrayAdapter.createFromResource(
                getActivity(),
                ResourcesManager.STRING_MAIN_STATISTIC_PERIOD,
                R.layout.spin_head_text_medium
        );

        adapterStatPeriod.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinPeriod.setAdapter(adapterStatPeriod);

        spinPeriod.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                if (spinPeriodNotInitSelectedItemCall) {
                    presenter.updateStatistic(i + 1);
                    sharedPrefManager.setHomePeriodPos(i);
                } else {
                    spinPeriodNotInitSelectedItemCall = true;
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

        spinPeriod.setSelection(sharedPrefManager.getHomePeriodPos());
    }

    private void setChartTypeSpinner() {
        spinChartType.setAdapter(new SpinIconTextHeadAdapter(
                getActivity(),
                R.layout.spin_head_icon_text_main_chart,
                R.id.tvSpinHeadIconTextMainChart,
                R.id.ivSpinHeadIconTextMainChart,
                R.layout.spin_drop_icon_text,
                R.id.tvSpinDropIconText,
                R.id.ivSpinDropIconText,
                resourcesManager.getStringArray(ResourcesManager.STRING_CHART_TYPE),
                resourcesManager.getIconArray(ResourcesManager.ICON_CHART_TYPE)));

        spinChartType.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                if (spinChartTypeNotInitSelectedItemCall) {
                    if (i == 1) {
                        chartStatistic.setVisibility(View.GONE);
                        if (presenter.isStatisticEmpty())
                            tvNoData.setVisibility(View.VISIBLE);
                        else {
                            chartStatisticPie.setVisibility(View.VISIBLE);
                            setStatisticPieChartData();
                        }
                    } else {
                        tvNoData.setVisibility(View.GONE);
                        chartStatisticPie.setVisibility(View.GONE);
                        chartStatistic.setVisibility(View.VISIBLE);
                        setStatisticBarChartData();
                    }
                    sharedPrefManager.setHomeChartTypePos(i);
                } else {
                    spinChartTypeNotInitSelectedItemCall = true;
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

        spinChartType.setSelection(sharedPrefManager.getHomeChartTypePos());
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(UpdateFrgHome event) {
        presenter.updateBalanceAndStatistic(spinPeriod.getSelectedItemPosition() + 1);
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(UpdateFrgHomeBalance event) {
        presenter.updateBalance();
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    public void onEvent(UpdateFrgHomeNewRates event) {
        presenter.updateRates();
        ratesInfoManager.prepareInfo();

        if (convert) {
            setBalance(spinBalanceCurrency.getSelectedItemPosition());
            presenter.updateTransactionStatisticArray(spinBalanceCurrency.getSelectedItemPosition());
            setStatisticSumTV();
            checkStatChartTypeForUpdate();
        }
    }

    private void setupCharts() {
        chartSetupManager.setupMainBarChart(chartBalance);
        chartSetupManager.setupMainBarChart(chartStatistic);
        chartSetupManager.setupMainPieChart(chartStatisticPie);
    }

    private void setBalanceBarChartData(double[] balance) {
        BarData data = presenter.getDataSetMainBalanceHorizontalBarChart(balance);
        data.setValueFormatter(new ChartLargeValueFormatter(!showOnlyIntegers));

        chartBalance.setData(data);
        chartBalance.animateXY(2000, 2000);
        chartBalance.invalidate();
    }

    private void setStatisticBarChartData() {
        BarData data = presenter.getDataSetMainStatisticHorizontalBarChart();
        data.setValueFormatter(new ChartLargeValueFormatter(!showOnlyIntegers));

        chartStatistic.setData(data);
        chartStatistic.animateXY(2000, 2000);
        chartStatistic.invalidate();
    }

    private void setStatisticPieChartData() {
        PieData data = presenter.getDataSetMainStatisticPieChart();
        data.setValueFormatter(new ChartLargeValueFormatter(!showOnlyIntegers));

        chartStatisticPie.setData(data);
        chartStatisticPie.animateXY(2000, 2000);
        chartStatisticPie.invalidate();
    }

    private void setStatisticSumTV() {
        tvStatisticSum.setText(presenter.getFormattedStatistic());
    }

    private void setBalanceTV(double[] balance) {
        tvBalanceSum.setText(presenter.getFormattedBalance(balance));
    }

    private void setBalance(int posCurrency) {
        double[] balance = presenter.getCurrentBalance(posCurrency);
        setBalanceTV(balance);
        setBalanceBarChartData(balance);
    }

    private void checkStatChartTypeForUpdate() {
        switch (spinChartType.getSelectedItemPosition()) {
            case 0:
                setStatisticBarChartData();
                break;
            case 1:
                if (presenter.isStatisticEmpty()) {
                    tvNoData.setVisibility(View.VISIBLE);
                    chartStatisticPie.setVisibility(View.GONE);
                } else {
                    tvNoData.setVisibility(View.GONE);
                    chartStatisticPie.setVisibility(View.VISIBLE);
                    setStatisticPieChartData();
                }
                break;
        }
    }

    @OnClick({R.id.ivMainBalanceSettings})
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.ivMainBalanceSettings:
                balanceSettingsDialog.show();
                break;
        }
    }


    @Override
    public void setBalanceAndStatistic(Pair<Map<String, double[]>, Map<String, double[]>> pair) {
        setBalance(spinBalanceCurrency.getSelectedItemPosition());
        setStatisticBarChartData();
        setStatisticSumTV();
        setChartTypeSpinner();
    }

    @Override
    public void updateBalanceAndStatistic(Pair<Map<String, double[]>, Map<String, double[]>> pair) {
        ratesInfoManager.prepareInfo();
        setBalance(spinBalanceCurrency.getSelectedItemPosition());
        setStatisticSumTV();
        checkStatChartTypeForUpdate();
    }

    @Override
    public void updateBalance(Map<String, double[]> map) {
        ratesInfoManager.prepareInfo();
        setBalance(spinBalanceCurrency.getSelectedItemPosition());
    }

    @Override
    public void updateStatistic(Map<String, double[]> map) {
        setStatisticSumTV();
        checkStatChartTypeForUpdate();
    }

    @Override
    public boolean isNeedToConvert() {
        return convert;
    }

    @Override
    public int getBalanceCurrencyPosition() {
        return spinBalanceCurrency.getSelectedItemPosition();
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/home/HomeFragment.kt
package com.androidcollider.easyfin.home

import android.os.Bundle
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemSelectedListener
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.getCustomView
import com.androidcollider.easyfin.R
import com.androidcollider.easyfin.common.app.App
import com.androidcollider.easyfin.common.events.*
import com.androidcollider.easyfin.common.managers.chart.setup.ChartSetupManager
import com.androidcollider.easyfin.common.managers.rates.rates_info.RatesInfoManager
import com.androidcollider.easyfin.common.managers.resources.ResourcesManager
import com.androidcollider.easyfin.common.managers.shared_pref.SharedPrefManager
import com.androidcollider.easyfin.common.managers.ui.dialog.DialogManager
import com.androidcollider.easyfin.common.ui.adapters.SpinIconTextHeadAdapter
import com.androidcollider.easyfin.common.ui.fragments.common.CommonFragmentWithEvents
import com.androidcollider.easyfin.common.utils.ChartLargeValueFormatter
import com.androidcollider.easyfin.common.utils.setSafeOnClickListener
import com.github.mikephil.charting.charts.HorizontalBarChart
import com.github.mikephil.charting.charts.PieChart
import com.google.android.material.button.MaterialButton
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import javax.inject.Inject

/**
 * @author Ihor Bilous
 */
class HomeFragment : CommonFragmentWithEvents(), HomeMVP.View {

    private lateinit var spinPeriod: Spinner
    private lateinit var spinBalanceCurrency: Spinner
    private lateinit var spinChartType: Spinner
    private lateinit var tvStatisticSum: TextView
    private lateinit var tvBalanceSum: TextView
    private lateinit var tvStatisticNoData: TextView
    private lateinit var tvBalanceNoData: TextView
    private lateinit var tvBalance: TextView
    private lateinit var ivBalanceSettings: MaterialButton
    private lateinit var chartStatistic: HorizontalBarChart
    private lateinit var chartBalance: HorizontalBarChart
    private lateinit var chartStatisticPie: PieChart
    private lateinit var chkBoxConvert: CheckBox
    private lateinit var chkBoxShowOnlyIntegers: CheckBox
    private lateinit var balanceSettingsDialog: MaterialDialog
    private var convert = false
    private var showOnlyIntegers = false

    // we can prevent init spinner issue with subscribing
    // to listeners in onRestoreInstanceState
    private var spinPeriodNotInitSelectedItemCall = false
    private var spinBalanceCurrencyNotInitSelectedItemCall = false
    //private var spinChartTypeNotInitSelectedItemCall = false

    @Inject
    lateinit var ratesInfoManager: RatesInfoManager

    @Inject
    lateinit var sharedPrefManager: SharedPrefManager

    @Inject
    lateinit var resourcesManager: ResourcesManager

    @Inject
    lateinit var chartSetupManager: ChartSetupManager

    @Inject
    lateinit var dialogManager: DialogManager

    @Inject
    lateinit var presenter: HomeMVP.Presenter

    override val contentView: Int
        get() = R.layout.frg_home

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (activity?.application as App).component?.inject(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setupUI(view)
        presenter.setView(this)
        presenter.loadBalanceAndStatistic(spinPeriod.selectedItemPosition + 1)
    }

    private fun setupUI(view: View) {
        spinPeriod = view.findViewById(R.id.spinMainPeriod)
        spinBalanceCurrency = view.findViewById(R.id.spinMainCurrency)
        spinChartType = view.findViewById(R.id.spinMainChart)
        tvStatisticSum = view.findViewById(R.id.tvMainStatisticSum)
        tvBalanceSum = view.findViewById(R.id.tvMainSumValue)
        tvStatisticNoData = view.findViewById(R.id.tvMainStatisticChartNoData)
        tvBalanceNoData = view.findViewById(R.id.tvMainBalanceChartNoData)
        tvBalance = view.findViewById(R.id.tvMainCurrentBalance)
        ivBalanceSettings = view.findViewById(R.id.btnMainBalanceSettings)
        chartStatistic = view.findViewById(R.id.chartHBarMainStatistic)
        chartBalance = view.findViewById(R.id.chartMainBalance)
        chartStatisticPie = view.findViewById(R.id.chartPieMainStatistic)

        ivBalanceSettings.setSafeOnClickListener { balanceSettingsDialog.show() }

        setupCharts()
        buildBalanceSettingsDialog()

        val balanceSettings = balanceSettingsDialog.getCustomView()
        chkBoxConvert = balanceSettings.findViewById(R.id.checkBoxMainBalanceSettingsConvert)
        chkBoxShowOnlyIntegers =
            balanceSettings.findViewById(R.id.checkBoxMainBalanceSettingsShowCents)
        convert = sharedPrefManager.mainBalanceSettingsConvertCheck
        showOnlyIntegers = sharedPrefManager.mainBalanceSettingsShowOnlyIntegersCheck
        chkBoxConvert.isChecked = convert
        chkBoxShowOnlyIntegers.isChecked = showOnlyIntegers
        chkBoxConvert.setOnCheckedChangeListener { _: CompoundButton?, b: Boolean ->
            sharedPrefManager.mainBalanceSettingsConvertCheck = b
            convert = b
            setBalance(spinBalanceCurrency.selectedItemPosition)
            presenter.updateTransactionStatisticArray(spinBalanceCurrency.selectedItemPosition)
            setStatisticSumTV()
            updateStatisticChartsSection()
        }
        chkBoxShowOnlyIntegers.setOnCheckedChangeListener { _: CompoundButton?, b: Boolean ->
            sharedPrefManager.mainBalanceSettingsShowOnlyIntegersCheck = b
            showOnlyIntegers = b
            setBalance(spinBalanceCurrency.selectedItemPosition)
            updateStatisticChartsSection()
        }
        ratesInfoManager.setupMultiTapListener(tvBalance, requireActivity())

        setBalanceCurrencySpinner()
        setStatisticPeriodSpinner()
    }

    private fun buildBalanceSettingsDialog() {
        activity?.let {
            balanceSettingsDialog = dialogManager.buildBalanceSettingsDialog(it)
        }
    }

    private fun setBalanceCurrencySpinner() {
        activity?.let {
            spinBalanceCurrency.adapter = SpinIconTextHeadAdapter(
                it,
                R.layout.spin_head_icon_text_main,
                R.id.tvSpinHeadIconTextMain,
                R.id.ivSpinHeadIconTextMain,
                R.layout.spin_drop_icon_text,
                R.id.tvSpinDropIconText,
                R.id.ivSpinDropIconText,
                resourcesManager.getStringArray(ResourcesManager.STRING_ACCOUNT_CURRENCY),
                resourcesManager.getIconArray(ResourcesManager.ICON_CURRENCY)
            )
            spinBalanceCurrency.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(
                    adapterView: AdapterView<*>?,
                    view: View?,
                    i: Int,
                    l: Long
                ) {
                    if (spinBalanceCurrencyNotInitSelectedItemCall) {
                        setBalance(i)
                        presenter.updateTransactionStatisticArray(i)
                        setStatisticSumTV()
                        updateStatisticChartsSection()
                        sharedPrefManager.homeBalanceCurrencyPos = i
                    } else {
                        spinBalanceCurrencyNotInitSelectedItemCall = true
                    }
                }

                override fun onNothingSelected(adapterView: AdapterView<*>?) {}
            }
            spinBalanceCurrency.setSelection(sharedPrefManager.homeBalanceCurrencyPos)
        }
    }

    private fun setStatisticPeriodSpinner() {
        activity?.let {
            val adapterStatPeriod: ArrayAdapter<*> = ArrayAdapter.createFromResource(
                it,
                ResourcesManager.STRING_MAIN_STATISTIC_PERIOD,
                R.layout.spin_head_text_medium
            )

            adapterStatPeriod.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            spinPeriod.adapter = adapterStatPeriod
            spinPeriod.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(
                    adapterView: AdapterView<*>?,
                    view: View?,
                    i: Int,
                    l: Long
                ) {
                    if (spinPeriodNotInitSelectedItemCall) {
                        presenter.updateStatistic(i + 1)
                        sharedPrefManager.homePeriodPos = i
                    } else {
                        spinPeriodNotInitSelectedItemCall = true
                    }
                }

                override fun onNothingSelected(adapterView: AdapterView<*>?) {}
            }
            spinPeriod.setSelection(sharedPrefManager.homePeriodPos)
        }
    }

    private fun setChartTypeSpinner() {
        activity?.let {
            spinChartType.adapter = SpinIconTextHeadAdapter(
                it,
                R.layout.spin_head_icon_text_main_chart,
                R.id.tvSpinHeadIconTextMainChart,
                R.id.ivSpinHeadIconTextMainChart,
                R.layout.spin_drop_icon_text,
                R.id.tvSpinDropIconText,
                R.id.ivSpinDropIconText,
                resourcesManager.getStringArray(ResourcesManager.STRING_CHART_TYPE),
                resourcesManager.getIconArray(ResourcesManager.ICON_CHART_TYPE)
            )
            spinChartType.onItemSelectedListener = object : OnItemSelectedListener {
                override fun onItemSelected(
                    adapterView: AdapterView<*>?, view: View?, i: Int, l: Long
                ) {
                    /*if (!spinChartTypeNotInitSelectedItemCall) {
                        spinChartTypeNotInitSelectedItemCall = true
                        return
                    }*/

                    if (presenter.isStatisticEmpty) {
                        tvStatisticNoData.visibility = View.VISIBLE
                        chartStatistic.visibility = View.GONE
                        chartStatisticPie.visibility = View.GONE
                    } else {
                        tvStatisticNoData.visibility = View.GONE
                        when (i) {
                            0 -> {
                                chartStatistic.visibility = View.VISIBLE
                                chartStatisticPie.visibility = View.GONE
                                setStatisticBarChartData()
                            }
                            1 -> {
                                chartStatistic.visibility = View.GONE
                                chartStatisticPie.visibility = View.VISIBLE
                                setStatisticPieChartData()
                            }
                        }
                    }
                    sharedPrefManager.homeChartTypePos = i
                }

                override fun onNothingSelected(adapterView: AdapterView<*>?) {}
            }
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: UpdateFrgHome) {
        presenter.updateBalanceAndStatistic(spinPeriod.selectedItemPosition + 1)
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: UpdateFrgHomeBalance) {
        presenter.updateBalance()
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: UpdateFrgHomeNewRates) {
        presenter.updateRates()
        ratesInfoManager.prepareInfo()
        if (convert) {
            setBalance(spinBalanceCurrency.selectedItemPosition)
            presenter.updateTransactionStatisticArray(spinBalanceCurrency.selectedItemPosition)
            setStatisticSumTV()
            updateStatisticChartsSection()
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN)
    fun onEvent(event: DBImported) {
        presenter.updateBalanceAndStatisticAfterDBImport(
            spinPeriod.selectedItemPosition + 1
        )
    }

    private fun setupCharts() {
        chartSetupManager.setupMainBarChart(chartBalance)
        chartSetupManager.setupMainBarChart(chartStatistic)
        chartSetupManager.setupMainPieChart(chartStatisticPie)
    }

    private fun setBalanceBarChartData(balance: DoubleArray) {
        val data = presenter.getDataSetMainBalanceHorizontalBarChart(balance, chartBalance)
        data.setValueFormatter(ChartLargeValueFormatter(!showOnlyIntegers))
        chartBalance.data = data
        chartBalance.animateXY(2000, 2000)
        chartBalance.invalidate()
    }

    private fun setStatisticBarChartData() {
        val data = presenter.getDataSetMainStatisticHorizontalBarChart(chartStatistic)
        data.setValueFormatter(ChartLargeValueFormatter(!showOnlyIntegers))
        chartStatistic.data = data
        chartStatistic.animateXY(2000, 2000)
        chartStatistic.invalidate()
    }

    private fun setStatisticPieChartData() {
        val data = presenter.getDataSetMainStatisticPieChart(chartStatisticPie)
        data.setValueFormatter(ChartLargeValueFormatter(!showOnlyIntegers))
        chartStatisticPie.data = data
        chartStatisticPie.animateXY(2000, 2000)
        chartStatisticPie.invalidate()
    }

    private fun setStatisticSumTV() {
        tvStatisticSum.text = presenter.formattedStatistic
    }

    private fun setBalanceTV(balance: DoubleArray) {
        tvBalanceSum.text = presenter.getFormattedBalance(balance)
    }

    private fun setBalance(posCurrency: Int) {
        val balance = presenter.getCurrentBalance(posCurrency)
        setBalanceTV(balance)
        updateBalanceChartSection(balance)
    }

    private fun updateBalanceChartSection(balance: DoubleArray) {
        if (presenter.isBalanceEmpty(balance)) {
            tvBalanceNoData.visibility = View.VISIBLE
            chartBalance.visibility = View.GONE
        } else {
            tvBalanceNoData.visibility = View.GONE
            chartBalance.visibility = View.VISIBLE
            setBalanceBarChartData(balance)
        }
    }

    private fun updateStatisticChartsSection() {
        if (presenter.isStatisticEmpty) {
            tvStatisticNoData.visibility = View.VISIBLE
            chartStatisticPie.visibility = View.GONE
            chartStatistic.visibility = View.GONE
        } else {
            tvStatisticNoData.visibility = View.GONE
            when (spinChartType.selectedItemPosition) {
                0 -> {
                    chartStatistic.visibility = View.VISIBLE
                    setStatisticBarChartData()
                }
                1 -> {
                    chartStatisticPie.visibility = View.VISIBLE
                    setStatisticPieChartData()
                }
            }
        }
    }

    override fun setBalanceAndStatistic() {
        setBalance(spinBalanceCurrency.selectedItemPosition)
        setStatisticBarChartData()
        setStatisticSumTV()
        setChartTypeSpinner()
    }

    override fun updateBalanceAndStatistic() {
        ratesInfoManager.prepareInfo()
        setBalance(spinBalanceCurrency.selectedItemPosition)
        setStatisticSumTV()
        updateStatisticChartsSection()
    }

    override fun updateBalanceAndStatisticAfterDBImport() {
        ratesInfoManager.prepareInfo()
        setBalance(spinBalanceCurrency.selectedItemPosition)
        setStatisticSumTV()
        updateStatisticChartsSection()
        EventBus.getDefault().post(UpdateFrgTransactions())
        EventBus.getDefault().post(UpdateFrgAccounts())
    }

    override fun updateBalance() {
        ratesInfoManager.prepareInfo()
        setBalance(spinBalanceCurrency.selectedItemPosition)
    }

    override fun updateStatistic() {
        setStatisticSumTV()
        updateStatisticChartsSection()
    }

    override val isNeedToConvert: Boolean
        get() = convert

    override val balanceCurrencyPosition: Int
        get() = spinBalanceCurrency.selectedItemPosition
}